### PR TITLE
RichText: Removing/merging richtext should only trigger if the selection is collapsed

### LINKED ISSUE #8638: Paragraph contents move to previous paragraph when deleted
**Describe the bug**

Sometimes attempting to delete the contents of a paragraph block sometimes will instead move the contents into the previous paragraph.

**To Reproduce**

1. Create a new post, switch to `Code Editor` mode, and paste in the following sample. Then, switch back to `Visual Mode`.

```
<!-- wp:paragraph -->
<p>paragraph one.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>paragraph two.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>paragraph three.</p>
<!-- /wp:paragraph -->
```

2. Highlight all of the text inside the block by clicking the mouse at the **end** of the text and dragging it to the beginning, by `Command+Shift+Left`, or by `Shift+PageUp`. _Don't_ use `Control+A`, or start at the beginning of the line, or use the `Delete` key, because the bug isn't triggered in those scenarios.

3. Hit the `Backspace` key to delete the paragraph.

**Expected behavior**

The contents of the block should be deleted. (Maybe the block itself should be deleted too, but I'm not really concerned about that.)

**Actual behavior**

Rather than the contents being deleted, they're moved into the previous paragraph, so that the raw document now looks like this:

```
<!-- wp:paragraph -->
<p>paragraph one.paragraph two.</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>paragraph three.</p>
<!-- /wp:paragraph -->
```

**Desktop (please complete the following information):**
 - OS X 10.13.6
 - Chrome 67.0.3396.99
 - Gutenberg `master` @ 4fd5c20af4a8a46c92048574e363b07977aa0f3f

### LINKED ISSUE #8639: Unable to use backspace to delete text content in Paragraph block when block is the first block, and text has been reverse-selected to the beginning
**Describe the bug**
This is an obscure one, but produces very unexpected behaviour.

When editing the first paragraph block in a post, if the user tries to delete some text by reverse-selecting to the start of the block, and then hitting backspace, nothing happens.

**To Reproduce**
1. Create a new post
2. Navigate to the first block and type some text
3. Reverse select the text to the beginning (Cmd+Shift+left arrow)
4. Hit backspace
5. Observe that the text was not deleted

**Expected behavior**
The selected text should have been deleted

**Desktop (please complete the following information):**
 - Tested on Chrome, Mac OS, but the problem is cross-browser

**Additional context**
- Gutenberg 3.4
- The issue is caused by the RichText component preventing propagation and default behaviour of keyboard events when the caret is at the start or end of a block:
https://github.com/WordPress/gutenberg/blob/98274f0e29eda553e038e2e6a7df564b502d039c/packages/editor/src/components/rich-text/index.js#L439-L463

The calls to `stopImmediatePropagation` and `preventDefault` should only be made when the block is about to be deleted. Here, because it's the first block, the backspace will not result in deletion.

### PR DESCRIPTION
**Pull request by youknowriad**
closes #8639
closes #8638

**Testing instructions**

Repeat the testing instructions from the issue, it should work as intended.

### CODE REVIEW

File: packages/editor/src/components/rich-text/index.js [hunk 1]
```diff
@@ -437,8 +437,10 @@ export class RichText extends Component {
 		const { keyCode } = event;
 
 		if (
-			( keyCode === BACKSPACE && isHorizontalEdge( rootNode, true ) ) ||
-			( keyCode === DELETE && isHorizontalEdge( rootNode, false ) )
+			window.getSelection().isCollapsed && (
```
Comment (iseulde) [inline packages/editor/src/components/rich-text/index.js]: Should we continue with setting it at the top as browser dependencies? `const { getSelection } = window;`

File: packages/editor/src/components/rich-text/index.js [hunk 2]
```diff
@@ -437,8 +437,10 @@ export class RichText extends Component {
...
+				( keyCode === BACKSPACE && isHorizontalEdge( rootNode, true ) ) ||
```
Comment (iseulde) [inline packages/editor/src/components/rich-text/index.js]: For some reason I would have thought that `isHorizontalEdge` also does the check, as navigating also requires it.

**Review by tofumatt — APPROVED** [review]
Tested locally, works, nice 👍

File: packages/editor/src/components/rich-text/index.js (hunk 2 above)
Comment (youknowriad) [reply packages/editor/src/components/rich-text/index.js]: Yes, I noticed it was also checked in the `WritingFlow`. I didn't want to change `isHorizontalEdge` behavior as I wasn't sure about the implications

File: test/e2e/specs/splitting-merging.test.js [hunk 3]
```diff
@@ -87,4 +87,18 @@ describe( 'splitting and merging blocks', () => {
 
 		expect( await getEditedPostContent() ).toMatchSnapshot();
 	} );
+
+	it( 'Should not merge paragraphs if the selection is not collapsed', async () => {
```
[the remark this answers is not in the document]
Comment (youknowriad) [reply test/e2e/specs/splitting-merging.test.js]: Starting today (if I remember :P)

File: packages/editor/src/components/rich-text/index.js (hunk 2 above)
Comment (aduth) [reply packages/editor/src/components/rich-text/index.js]: Maybe. As it reads, I see no reason why `isHorizontalEdge` should have any concern about whether the selection is collapsed or not. This assumes there's a use case for wanting to know if there's an uncollapsed selection at the horizontal edge. We don't have this need in our project, which might bias our opinion on it. Or maybe "is horizontal edge" is meant to mean that pressing ArrowLeft/ArrowRight would have no effect (in a traditional input).

**Comment by aduth** [comment]
This duplicated effort from #8306

**Pull request merged**